### DISCUSSION TITLE
Regarding the avoidance of N+1 queries when using the combination of `prismaConnection` and `t.relation`

### DISCUSSION THREAD

**Discussion by touyu** [hayes/pothos#926] [Q&A] [answered]
As shown in the code below, when `prismaField` is used in `queryField` and `t.relation` is used inside `prismaObject`, a query is issued that avoids the N+1 problem.

```typescript
builder.prismaObject('Article', {
  fields: (t) => ({
    databaseId: t.exposeInt('databaseId'),
    creator: t.relation('creator', {
      nullable: true,
    }),
  }),
})

builder.queryField('articles', (t) =>
  t.prismaField({
    type: ['Article'],
    resolve: (query) => {
      return prisma.article.findMany({
        ...query,
      })
    },
  })
)
```

Next, As shown below, when `prismaConnection` is used in `queryField` and `t.relation` is used inside `prismaNode`, a large number of queries are issued.

In such a case, doesn't `t.relation` generate a query that avoids the N+1 problem?

```typescript
builder.prismaNode('Article', {
  id: { field: 'databaseId' },
  fields: (t) => ({
    databaseId: t.exposeInt('databaseId'),
    creator: t.relation('creator', {
      nullable: true,
    }),
  }),
})

builder.queryField('articles', (t) =>
  t.prismaConnection({
    type: 'Article',
    cursor: 'databaseId',
    resolve: (query, parent, args) => {
      return prisma.article.findMany({
        ...query,
      })
    },
  })
)
```

**Comment by hayes** [comment]
I would expect that this would create one Prisma query, and which would result in 2 SQL queries.  Do you have any more details to share on what's going on here? Wasn't able to reproduce the issue, and there are several tests covering similar patterns showing optimized queries

**Comment by touyu** [comment]
@hayes 
Thanks for the reply.
I created a sample code that can be reproduced here.
Could you please check here?
https://github.com/touyu/pothos-prisma-query

README.md describes the query to reproduce.

**Reply by touyu** [reply]
I may have noticed one possible cause of the problem.
A GraphQL Query containing `nodes` may be causing this.
```graphql
{
  articles(first: 30) {
    nodes {
      databaseId
      creator {
        databaseId
      }
    }
  }
}
```

`edges`, there was no problem.
```graphql
{
  articles(first: 30) {
    edges {
      node {
        databaseId
        creator {
          databaseId
        }
      }
    }
  }
}
```

**Reply by hayes** [reply]
@touyu this looks like a combination of 2 issues.

The first issue is related to an issue with prisma's batching not correctly being applied to findUniqueOrThrow (https://github.com/prisma/prisma/issues/18838)

The second issue is the node's field uses a fairly simple implementation that doesn't hoist nested dependencies, so they are loaded in a less efficient way (would normally 2 queries not n+1 if prisma batching was applied)

**Reply by touyu** [reply]
I Understand. Thank you.
I will keep watch on the Prisma's issue that you shared with me.

**Reply by hayes** [reply]
@touyu are you using the jsonprotocol preview feature for prisma?

**Reply by hayes** [reply]
opened a couple PRs to prisma to hopefully get this fixed: https://github.com/prisma/prisma-engines/pull/4014 and https://github.com/prisma/prisma/pull/19554)

**Reply by touyu** [reply]
> are you using the jsonprotocol preview feature for prisma?

I don't use jsonprotocol. 

> opened a couple PRs to prisma to hopefully get this fixed: https://github.com/prisma/prisma-engines/pull/4014 and https://github.com/prisma/prisma/pull/19554)

Great! I will keep watch on these PRs.
Thank you!

**Reply by hayes** [reply, marked as answer]
can you upgrade prisma and pothos/plugin-prisma to the latest versions and see if your issue is resolved?  I pushed a fix for optimizing nodes last week, and prisma has released my fix for optimizing findUniqueOrThrow calls, so I think this should work as expected now

**Reply by touyu** [reply]
@hayes Thank you for the great job you do!
